pkg/cmd/cli/cmd/importimage.go

```diff
@@ -67,45 +68,147 @@ func RunImportImage(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, arg
+		if len(from) == 0 && !confirm {
+			return fmt.Errorf("no image stream named %q exists, pass --confirm to create and import", name)
 		}
-		stream = &imageapi.ImageStream{
-			ObjectMeta: kapi.ObjectMeta{Name: streamName},
-			Spec:       imageapi.ImageStreamSpec{DockerImageRepository: from},
+
+		if len(ref.Tag) == 0 && all {
+			stream = &imageapi.ImageStream{
+				ObjectMeta: kapi.ObjectMeta{Name: name},
+				Spec:       imageapi.ImageStreamSpec{DockerImageRepository: from},
+			}
+		} else {
+			stream = &imageapi.ImageStream{
+				ObjectMeta: kapi.ObjectMeta{Name: name},
+				Spec: imageapi.ImageStreamSpec{
+					Tags: map[string]imageapi.TagReference{
+						ref.Tag: {
+							From: &kapi.ObjectReference{
+								Kind: "DockerImage",
+								Name: from,
+							},
+						},
+					},
+				},
+			}
 		}
+
 	} else {
 		if len(stream.Spec.DockerImageRepository) == 0 && len(stream.Spec.Tags) == 0 {
 			return fmt.Errorf("image stream has not defined anything to import")
 		}
-		if len(from) != 0 {
-			if from != stream.Spec.DockerImageRepository {
+		if len(ref.Tag) == 0 && all {
+			if len(from) != 0 && from != stream.Spec.DockerImageRepository {
 				if !confirm {
+					if len(stream.Spec.DockerImageRepository) == 0 {
+						return fmt.Errorf("the image stream does not currently import an entire Docker repository, pass --confirm to update")
+					}
 					return fmt.Errorf("the image stream has a different import spec %q, pass --confirm to update", stream.Spec.DockerImageRepository)
 				}
 				stream.Spec.DockerImageRepository = from
 			}
+		} else {
+			var tag imageapi.TagReference
+			if existing, ok := stream.Spec.Tags[ref.Tag]; ok {
+				tag = existing
+				delete(tag.Annotations, imageapi.DockerImageRepositoryCheckAnnotation)
+			} else {
+				tag = imageapi.TagReference{
+					From: &kapi.ObjectReference{
+						Kind: "DockerImage",
+						Name: from,
+					},
+				}
+			}
+			stream.Spec.Tags[ref.Tag] = tag
 		}
 	}
 
-	if stream.Annotations != nil {
-		delete(stream.Annotations, imageapi.DockerImageRepositoryCheckAnnotation)
+	// Attempt the new, direct import path
+	isi := &imageapi.ImageStreamImport{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:            stream.Name,
+			Namespace:       namespace,
+			ResourceVersion: stream.ResourceVersion,
+		},
+		Spec: imageapi.ImageStreamImportSpec{Import: true},
+	}
+	if all {
+		isi.Spec.Repository = &imageapi.RepositoryImportSpec{
+			From: kapi.ObjectReference{
+				Kind: "DockerImage",
+				Name: from,
+			},
+			ImportPolicy: imageapi.TagImportPolicy{Insecure: insecure},
+		}
 	} else {
-		stream.Annotations = make(map[string]string)
+		isi.Spec.Images = append(isi.Spec.Images, imageapi.ImageImportSpec{
+			From: kapi.ObjectReference{
+				Kind: "DockerImage",
+				Name: from,
+			},
+			To:           &kapi.LocalObjectReference{Name: ref.Tag},
+			ImportPolicy: imageapi.TagImportPolicy{Insecure: insecure},
+		})
 	}
 
+	// TODO: add dry-run
+	_, err = imageStreamClient.Import(isi)
```

Comment:
Eh, this call should be fast because it's direct and we limit tags - the old call was slow (because import of mysql could take 5-10 minutes while we downloaded hundreds of tags, and it was also gated by the import controller).  We won't take longer than it takes to find ~5 tags from a remote registry, and CTRL+C is implicit in waiting.